File: friendship-service/friendship-service.py
class FriendshipService:

    def __init__(self):
        # initialize your data structure here.
        self.followers = dict()
        self.followings = dict()

    # @param {int} user_id
    # return {int[]} all followers and sort by user_id
    def get_followers(self, user_id):
        # Write your code here
        if user_id in self.followers:
            results = list(self.followers[user_id])
            results.sort()
            return results
        return []

    # @param {int} user_id
    # return {int[]} all followers and sort by user_id
    def get_followings(self, user_id):
        # Write your code here
        if user_id in self.followings:
            results = list(self.followings[user_id])
            results.sort()
            return results
        return []

    # @param {int} from_user_id
    # @param {int} to_user_id
    # from_user_id follows to_user_id
    def follow(self, from_user_id, to_user_id):
        # Write your code here
        if from_user_id not in self.followers:
            self.followers[from_user_id] = set()
        self.followers[from_user_id].add(to_user_id)
        if to_user_id not in self.followings:
            self.followings[to_user_id] = set()
        self.followings[to_user_id].add(from_user_id)

    # @param {int} from_user_id
    # @param {int} to_user_id
    # from_user_id unfollows to_user_id
    def unfollow(self, from_user_id, to_user_id):
        # Write your code here
        if from_user_id in self.followers and to_user_id in self.followers[from_user_id]:
            self.followers[from_user_id].remove(to_user_id)
        if to_user_id in self.followings and from_user_id in self.followings[to_user_id]:
            self.followings[to_user_id].remove(from_user_id)

File: friendship-service/friendship-service.cc
class FriendshipService {
private:
    map<int, vector<int> > from_to;
    map<int, vector<int> > to_from;

public:
    FriendshipService() {
        // initialize your data structure here.
        from_to.clear();
        to_from.clear();
    }

    // @param user_id an integer
    // return all followers and sort by user_id
    vector<int>  getFollowers(int user_id) {
        // Write your code here
        vector<int> &tmp = from_to[user_id];
        sort(tmp.begin(), tmp.end());
        return tmp;
    }

    // @param user_id an integer
    // return all followings and sort by user_id
    vector<int>  getFollowings(int user_id) {
        // Write your code here
        vector<int> &tmp = to_from[user_id];
        sort(tmp.begin(), tmp.end());
        return tmp;
    }

    // @param from_user_id an integer
    // @param to_user_id an integer
    // from user_id follows to_user_id
    void follow(int from_user_id, int to_user_id) {
        // Write your code here
        vector<int> &from = from_to[from_user_id];
        vector<int> &to = to_from[to_user_id];
        vector<int>::iterator from_it = find(from.begin(),from.end(),to_user_id);
        if(from_it == from.end()) {
            from.push_back(to_user_id);
            to.push_back(from_user_id);
        }
    }

    // @param from_user_id an integer
    // @param to_user_id an integer
    // from user_id unfollows to_user_id
    void unfollow(int from_user_id, int to_user_id) {
        // Write your code here
        vector<int> &from = from_to[from_user_id];
        vector<int> &to = to_from[to_user_id];
        vector<int>::iterator from_it = find(from.begin(),from.end(),to_user_id);
        if(from_it != from.end()) {
            from.erase(from_it);
            vector<int>::iterator to_it = find(to.begin(),to.end(),from_user_id);
            to.erase(to_it);
        }
    }
};
